File: vnshop_mobile/lib/features/reviews/domain/entities/review_summary.dart
import 'package:equatable/equatable.dart';

import 'review.dart';

class ReviewSummary extends Equatable {
  const ReviewSummary({
    required this.average,
    required this.count,
    required this.distribution,
  });

  static const empty = ReviewSummary(
    average: 0,
    count: 0,
    distribution: {1: 0, 2: 0, 3: 0, 4: 0, 5: 0},
  );

  factory ReviewSummary.fromReviews(Iterable<Review> reviews) {
    final distribution = <int, int>{1: 0, 2: 0, 3: 0, 4: 0, 5: 0};
    var total = 0;
    var count = 0;

    for (final review in reviews) {
      final rating = review.rating.clamp(1, 5);
      distribution[rating] = distribution[rating]! + 1;
      total += rating;
      count++;
    }

    return ReviewSummary(
      average: count == 0 ? 0 : (total / count * 10).round() / 10,
      count: count,
      distribution: Map.unmodifiable(distribution),
    );
  }

  final double average;
  final int count;
  final Map<int, int> distribution;

  @override
  List<Object?> get props => [average, count, distribution];
}

File: fe/src/features/reviews/model/review-summary.ts
export type ReviewRating = 1 | 2 | 3 | 4 | 5;

export interface ReviewSummary {
  average: number;
  count: number;
  distribution: Record<ReviewRating, number>;
}

interface RatingReview {
  rating: number;
}

export function summarizeReviews(reviews: readonly RatingReview[]): ReviewSummary {
  const distribution: Record<ReviewRating, number> = {
    1: 0,
    2: 0,
    3: 0,
    4: 0,
    5: 0,
  };

  const normalizedRatings = reviews.map(({ rating }) => {
    const normalized = Math.min(5, Math.max(1, Math.round(rating))) as ReviewRating;
    distribution[normalized] += 1;
    return normalized;
  });

  const count = normalizedRatings.length;
  const total = normalizedRatings.reduce((sum, rating) => sum + rating, 0);

  return {
    average: count === 0 ? 0 : Math.round((total / count) * 10) / 10,
    count,
    distribution,
  };
}
